### PR TITLE
Allow Gubernator to read jUnit from all XML in a bucket

### PR DESCRIPTION
**Pull request by stevekuznetsov**
One of the most onerous requirements for the GCS layout is that all
jUnit need to be in a certain place and fit a certain file name format.
This patch removes that requirement and will show jUnit output from all
jUnit XML in the test bucket.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

/assign @rmmh

### CODE REVIEW

**Comment by rmmh** [comment]
Can you improve the parser to reject non-junit XMLs without showing errors? It's possible artifacts will have XML files conforming to different schemas.

**Comment by stevekuznetsov** [comment]
@rmmh sounds good, will do

File: gubernator/view_base.py [hunk 1]
```diff
@@ -120,6 +120,18 @@ def gcs_ls(path):
         path += '/'
     return list(gcs.listbucket(path, delimiter='/'))
 
+@memcache_memoize('gs-ls-recursive://', expires=60)
+def gcs_ls_recursive(path):
+    """Enumerate files in a GCS directory recursively. Returns a list of FileStats."""
+    if path[-1] != '/':
+        path += '/'
+
+    files = list(gcs.listbucket(path, delimiter='/'))
```
Comment (rmmh) [inline gubernator/view_base.py]: This doesn't have to be recursive. Remove the delimiter and you'll get a listing of *all* files. GCS doesn't have directories, just k/v pairs!

Comment (stevekuznetsov) [reply gubernator/view_base.py]: I had tried that but I thought I got the entire bucket, not just what was under the path. I'll change to using that functionality, though, as it's probably faster.

File: gubernator/view_build.py [hunk 2]
```diff
@@ -71,8 +71,9 @@ def parse_xml(self, xml, filename):
             try:
                 tree = ET.fromstring(re.sub(r'[\x00\x80-\xFF]+', '?', xml))
             except ET.ParseError, e:
-                self.failed.append(
-                    ('Gubernator Internal Fatal XML Parse Error', 0.0, str(e), filename, ''))
+                if re.match(r'junit.*\.xml', os.path.basename(filename)):
```
Comment (stevekuznetsov) [inline gubernator/view_build.py]: @rmmh left the errors in this case so we don't lose that signal for existing jUnit producers

Comment (rmmh) [reply gubernator/view_build.py]: can you do the same for the else case below? `logging.error('unable to find failures, unexpected tag %s', tree.tag)`

Comment (stevekuznetsov) [reply gubernator/view_build.py]: Ah yes sorry, missed that one

File: gubernator/view_build.py [hunk 3]
```diff
@@ -138,8 +139,8 @@ def build_details(build_dir):
     started = json.loads(started)
     finished = json.loads(finished)
 
-    junit_paths = [f.filename for f in view_base.gcs_ls('%s/artifacts' % build_dir)
-                   if re.match(r'junit_.*\.xml', os.path.basename(f.filename))]
+    junit_paths = [f.filename for f in view_base.gcs_ls_recursive('%s/artifacts' % build_dir)
+                   if re.match(r'.*\.xml', os.path.basename(f.filename))]
```
Comment (rmmh) [inline gubernator/view_build.py]: could be simplified to `if f.filename.endswith('.xml')`

**Pull request merged**